crates/ruma-identifiers/src/transaction_id.rs

```diff
@@ -0,0 +1,36 @@
+/// A Matrix transaction ID.
+///
+/// Transaction IDs in Matrix are opaque strings. This type is provided simply for its semantic
+/// value.
+///
+/// You can create one from a string (using `.into()`) but the recommended way is to use
+/// `TransactionId::new()` to generate a random one. If that function is not available for you, you
+/// need to activate this crate's `rand` Cargo feature.
+#[repr(transparent)]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct TransactionId(str);
+
+impl TransactionId {
+    /// Creates a random transaction ID.
+    ///
+    /// This will currently be a UUID without hyphens, but no guarantees are made about the
+    /// structure of transaction IDs generated from this function.
+    #[cfg(feature = "rand")]
+    pub fn new() -> Box<Self> {
+        let id = uuid::Uuid::new_v4();
+
+        let mut buf: Box<[u8]> = Box::new([0; 32]);
+        id.to_simple().encode_lower(&mut buf);
+
+        // SAFETY: str has the same representation as [u8], as guaranteed by
+        // `str::into_boxed_bytes`¹ saying "without copying or allocating".
+        //
+        // ¹ https://doc.rust-lang.org/std/primitive.str.html#method.into_boxed_bytes
+        let raw = Box::into_raw(buf) as *mut str;
+        let buf_str: Box<str> = unsafe { Box::from_raw(raw) };
```

Comment:
`str` also has to be valid UTF8, you also have to uphold that guarantee.

Reply:
Yes, UUIDs are always valid UTF-8. I suppose I should update the comment.